Delete unused top level binding code action

serhiip opened the pull request:
This is my first contribution. My goal is contribute to haskell-language-server / ghcide constantly. This adds a code action that deletes unused top level binding.

Please, let me know how this should be improved.

This closes #54 (I think)

digitalasset-cla commented:
[![CLA assistant check](https://cla.digitalasset.com/pull/badge/not_signed)](https://cla.digitalasset.com/digital-asset/ghcide?pullRequest=657) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.digitalasset.com/digital-asset/ghcide?pullRequest=657) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.digitalasset.com/check/digital-asset/ghcide?pullRequest=657) it.</sub>

ndmitchell commented on test/exe/Main.hs:
```diff
@@ -1144,6 +1145,40 @@ insertNewDefinitionTests = testGroup "insert new definition actions"
+        "{-# OPTIONS_GHC -Wunused-top-binds #-}"
+        , "module A (some) where"
+        , ""
+        , ""
```
A very minor remark, but beforehand they had 1-line gaps, now they have 2-line gaps. Should we expand one line after if it's blank? Or maybe this is getting ahead of ourselves and there is no need.

cocreature replied:
Makes sense to me. What do you think @serhiip?

serhiip replied:
For sure expanding the text edit until the next binding in a file is a good thing. I was thinking about this initially but only after your comment I've figured how to actually do it: now the code works by finding the next binding after current one and expands the deletion until that next binding (if any). To make that happen I rearranged my code a bit. Hope it makes more sense now. Thanks for suggestion

cocreature commented on src/Development/IDE/Plugin/CodeAction.hs:
```diff
@@ -180,6 +181,24 @@ suggestRemoveRedundantImport ParsedModule{pm_parsed_source = L _  HsModule{hsmod
+            = [("Delete ‘" <> name <> "’", [TextEdit (nextLine . srcSpanToRange $ l) "" | (L l _) <- sameName])]
+    | otherwise = []
+    where
+      isTopLevel l = (_character . _start) (srcSpanToRange l) == 0
```
You are basically duplicating this logic in line 188. How about we change this to accept a range and then we can use it for both cases (composing with `srcSpanToRange` at one call site).

serhiip replied:
Indeed! There was a bug with infix bindings, so I've eradicated this line to fix that bug

serhiip commented:
> This looks really good. Thanks for your contribution!
> 
> My guess is this doesn't work for operators, or functions defined like `` x `f` y = ... ``? Not a problem, as they are vastly more rare, and this is a code action.

Good point. And actually there was a bug :). Luckily it was easy to fix. I pushed the fixed version and added corresponding test (just in case). It was a one line fix so I've incorporated it together with other changes I did.

cocreature commented on src/Development/IDE/Plugin/CodeAction.hs:
```diff
@@ -180,6 +181,30 @@ suggestRemoveRedundantImport ParsedModule{pm_parsed_source = L _  HsModule{hsmod
+      forwardLines lines r = r {_end = (_end r) {_line = (_line . _end $ r) + lines, _character = 0}}
+
+      toNextBinding bindings r@Range { _end = Position {_line = l} }
+        | Just (Range { _start = Position {_line = l'}}, _) <- find ((> l) . _line . _start . fst) bindings
```
Is `hsmodDecls` guaranteed to be sorted by ranges? Looking at haddock’s `collectDocs` it always seems to be called after `sortLocated` suggesting that we probably need this here as well.

serhiip replied:
Order in `hsmodDecls` doesn't matter really - I use `find` to locate next binding within all the top level bindings. I might have missed the point why `hsmodDecls` must be sorted

cocreature replied:
`find` will give you the first match. where the condition is true, consider what happens if you have 3 bindings. You want to delete the first. Assume `hsModDecls` has the order `1, 3, 2`. Then your `find` here will find `3` and you will end up deleting `2` as well

serhiip replied:
My bad. That is indeed an issue - I've pushed an update. Thanks for catching this
